### PR TITLE
Live casting system

### LINKED ISSUE #984: Live cast system [$205]
# Live Cast System

twifysoft.net has a very good cast system, but I understand if they wish to keep the sources closed. 
However, I believe a good live cast system will help the growth of this community, and motivate more people to jump on projects based on this repository. 

**So I want a live cast system, on this official repository.**
## How I want it to work (copied from twifysoft):
- You can start casting by typing the command !cast ingame. 
- You can also start casting with a password, by using the command !cast password. 
- Login to the server without account name to watch a live cast. 
- If the cast has a password, enter the password in the password login field (still empty account name).
- An in-game cast group chat channel. If the caster has the channel open, it is considered enabled, if caster is not in the channel its considered disabled. 
## Code quality guidelines:

[@marksamman](https://github.com/otland/forgottenserver/issues/294#issuecomment-30718539):
- The use of the dispatcher thread must be minimal (viewers should receive a copy of the packets sent to the caster in asio thread). 
- All caster variables must be stored in (or in another object referenced from) the ProtocolGame instance of the caster player.

I will be the first here to take initiative and use Bountysource, I hope more people will follow. :+1: 
## <bountysource-plugin>

There is a **[$205 open bounty](https://www.bountysource.com/issues/4310302-live-cast-system-145?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by slavidodo**
Continues #994 . And closes #984 
Additional features than the ones mentioned there in the issue are that you can mute/ban a specific spectator and pause the cast while in need to do.
Furthermore, containers, channels and fight modes are synced. Accessing the castinfo should ONLY be in the dispatcher thread.

### CODE REVIEW

File: src/protocolspectator.cpp [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+	casterProtocol->sendChannelMessage(name, text, TALKTYPE_CHANNEL_Y, CHANNEL_CAST);
+	lastSpeak = OTSYS_TIME();
+}
+
+void ProtocolSpectator::parseCloseChannel(NetworkMessage& msg)
+{
+	uint16_t channelId = msg.get<uint16_t>();
+	if (channelId == CHANNEL_CAST) {
+		sendCastChannel();
```
Comment (djarek) [inline src/protocolspectator.cpp]: You have a data race on the caster's castinfo here.

File: data/migrations/20.lua [hunk 2]
```diff
@@ -0,0 +1,3 @@
+﻿function onUpdateDatabase()
+	return false
+end
```
Comment (brunominervino) [inline data/migrations/20.lua]: EOF

File: data/talkactions/scripts/livecast.lua [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	elseif param == "off" then
+		if player:isLiveCasting() then
+			player:stopLiveCasting()
+		end
+	else -- a password
+		if not player:isLiveCasting() then
+			player:startLiveCasting(defaultParam)
+		end
+	end
+end
```
Comment (brunominervino) [inline data/talkactions/scripts/livecast.lua]: EOF

File: src/protocolgame.cpp [hunk 4]
```diff
@@ -1116,1937 +926,198 @@ void ProtocolGame::parseSeekInContainer(NetworkMessage& msg)
...
-{
-	uint8_t opcode = msg.getByte();
-	const std::string& buffer = msg.getString();
-
-	// process additional opcodes via lua script event
-	addGameTask(&Game::parsePlayerExtendedOpcode, player->getID(), opcode, buffer);
-}
+	sendChannelMessage("[Cast system]", "There is no spectator with that name.", TALKTYPE_CHANNEL_R1, CHANNEL_CAST, false);
+	return false;
+}
```
Comment (brunominervino) [inline src/protocolgame.cpp]: EOF

File: src/protocolspectator.cpp [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+}
+
+void ProtocolSpectator::reOpenCastChannel()
+{
+	// dispatcher thread
+	if (player && casterProtocol && player->client == casterProtocol->shared_from_this()) {
+		sendCastChannel();
+		sendChannelMessage("[Cast System]", "You can't close this channel!", TALKTYPE_CHANNEL_R1, CHANNEL_CAST);
+	}
+}
```
Comment (brunominervino) [inline src/protocolspectator.cpp]: EOF

File: src/protocolspectator.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+		friend class ProtocolGame;
+
+		uint32_t lastSpeak = 0;
+
+		std::string name; // spectator name
+		ProtocolGame* casterProtocol = nullptr;
+};
+
+#endif
```
Comment (brunominervino) [inline src/protocolspectator.h]: EOF

File: src/protocolgamebase.cpp [hunk 7]
```diff
@@ -0,0 +1,2189 @@
...
+		msg.addByte(serverFluidToClient(item.subType));
+	} else {
+		msg.addByte(0x00);
+	}
+
+	msg.addString(item.realName);
+	msg.add<uint32_t>(it.weight);
+	msg.add<uint32_t>(item.buyPrice);
+	msg.add<uint32_t>(item.sellPrice);
+}
```
Comment (brunominervino) [inline src/protocolgamebase.cpp]: EOF

File: src/iologindata.cpp [hunk 8]
```diff
@@ -183,16 +183,37 @@ void IOLoginData::updateOnlineStatus(uint32_t guid, bool login)
 	if (g_config.getBoolean(ConfigManager::ALLOW_CLONES)) {
 		return;
 	}
-
+
```
Comment (brunominervino) [inline src/iologindata.cpp]: Trailing Space

File: data/talkactions/scripts/livecastcommands.lua [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+	elseif words == "!unban" then
+		player:unBanCastSpectator(param)
+	elseif words == "!mute" then
+		player:muteCastSpectator(param)
+	elseif words == "!unmute" then
+		player:unMuteCastSpectator(param)
+	end
+	
+	return false
+end
```
Comment (brunominervino) [inline data/talkactions/scripts/livecastcommands.lua]: EOF

File: data/migrations/20.lua (hunk 2 above)
Comment (slavidodo) [inline data/migrations/20.lua]: Thanks, I'd like you to make a pr (in my repo) fixing thus as I am not able now.

**Comment by brunominervino** [comment]
My mistake to remove my reviews.

**Comment by thexamx** [comment]
For some reason this cast system does not update the spectators in database
https://i.gyazo.com/74102a3c122adc58f9f15f23ef5062fe.png
It is always on 0.

Can someone reproduce that?

**Comment by slavidodo** [comment]
Thanks Themax!

File: src/configmanager.cpp [hunk 10]
```diff
@@ -111,6 +111,7 @@ bool ConfigManager::load()
 
 		integer[SQL_PORT] = getGlobalNumber(L, "mysqlPort", 3306);
 		integer[GAME_PORT] = getGlobalNumber(L, "gameProtocolPort", 7172);
+		integer[CAST_PORT] = getGlobalNumber(L, "loveCastProtocolPort", 7173);
```
Comment (vankk) [inline src/configmanager.cpp]: Are you in love?

Comment (thexamx) [inline src/configmanager.cpp]: can you teach me how do you do comment like this ?

Comment (WibbenZ) [inline src/configmanager.cpp]: Files changes -> go to a line, put the marker next to a line number and click the blue + box :)

File: src/iologindata.cpp [hunk 11]
```diff
@@ -186,13 +186,36 @@ void IOLoginData::updateOnlineStatus(uint32_t guid, bool login)
...
+	std::ostringstream query;
+	query << "UPDATE `players_online` set `cast_on` = 1, `cast_password` = " << db.escapeString(password) << ", `cast_spectators` = 0 WHERE `player_id` = " << guid;
+	db.executeQuery(query.str());
+}
+
+void IOLoginData::updateCast(uint32_t guid, uint32_t spectators)
+{
+	Database& db = Database::getInstance();
+	std::ostringstream query;
+	query << "UPDATE `players_online` set `cast_spectators = " << spectators << " WHERE `player_id` = " << guid;
```
Comment (thexamx) [inline src/iologindata.cpp]: query << "UPDATE `players_online` set `cast_spectators` = " << spectators << " WHERE `player_id` = " << guid;

You are missing "`"

File: src/configmanager.cpp (hunk 10 above)
Comment (thexamx) [inline src/configmanager.cpp]: Thank you

File: src/iologindata.cpp (hunk 11 above)
Comment (slavidodo) [inline src/iologindata.cpp]: Can u make a pr, I can't have a pc for a while. Thanks.

Comment (thexamx) [inline src/iologindata.cpp]: Done, if I did this correct

**Comment by westwol** [comment]
There's a crash issue.

gdb:
https://pastebin.com/vFVxLq3L

**How to reproduce:**
- Start two casts with different players.
- Logout them almost at the same time.

**Comment by slavidodo** [comment]
@ranisalt @djarek

**Comment by infister** [comment]
The crash issue

> There's a crash issue.
> 
> gdb:
> https://pastebin.com/vFVxLq3L
> 
> How to reproduce:
> 
> Start two casts with different players.
> Logout them almost at the same time.

I have probably fixed it by copying kick functions from ProtocolGame::release() to logout(). I observe that release() function is not casted instantly on logout, but on next ProtocolGame task, such as login...

**Comment by thexamx** [comment]
Actually I dont know if the bug is same as this previous but.
Open cast, login to cast with other client, logout caster, after few seconds server crashes with this gdb:
https://pastebin.com/raw/4QSRfR1D

**Comment by infister** [comment]
In my server this issue is resolved by doing actions I have described one post upper. Tested in many different ways. Before server was crashing, so there was huge improvement.

**Comment by westwol** [comment]
@infister can you make a pull request of this. Thanks in advance.

**Comment by DSpeichert** [comment]
@slavidodo we'll need a rebase of this.

File: src/otserv.cpp [hunk 12]
```diff
@@ -234,6 +235,7 @@ void mainLoader(int, char*[], ServiceManager* services)
 
 	// Game client protocols
 	services->add<ProtocolGame>(g_config.getNumber(ConfigManager::GAME_PORT));
+	services->add<ProtocolSpectator>(g_config.getNumber(ConfigManager::CAST_PORT));
```
Comment (comedinha) [inline src/otserv.cpp]: check if cast are enabled?

File: src/protocolgame.cpp [hunk 13]
```diff
@@ -335,45 +350,24 @@ void ProtocolGame::onRecvFirstMessage(NetworkMessage& msg)
...
+void ProtocolGame::writeToSpectatorsOutputBuffer(const NetworkMessage& msg)
 {
-	auto output = OutputMessagePool::getOutputMessage();
-	output->addByte(0x14);
-	output->addString(message);
-	send(output);
-	disconnect();
+	// dispatcher thread
+	if (castinfo.spectators.size() > 0) {
+		for(auto spectator : castinfo.spectators) {
```
Comment (comedinha) [inline src/protocolgame.cpp]: Space

**Comment by slavidodo** [comment]
I think I will wait until TFS 2 is released, then I will rebase this repo and resolve those conflicts.

**Comment by DSpeichert** [comment]
TFS 2.0 would be for breaking changes. It's still a long way till that. Since this is a new feature, it should not be incompatible and could probably fit right after TFS 1.3.

**Comment by slavidodo** [comment]
I have no intention to continue this. This should be closed by now.

File: src/otserv.cpp (hunk 12 above)
Comment (tysonstrange) [reply src/otserv.cpp:238]:
```
	if (g_config.getBoolean(ConfigManager::ENABLE_LIVE_CASTING)) {
		ProtocolGame::clearLiveCastInfo();
		services->add<ProtocolSpectator>(g_config.getNumber(ConfigManager::LIVE_CAST_PORT));
	}
```

**Review by pasturryx — COMMENTED** [review]
x

**Pull request closed without merging**